javascript-client/src/client.ts

```diff
@@ -47,12 +47,18 @@ export interface ProgressEvent {
 type ProgressCallback = (progress: ProgressEvent) => void;
 
 export default class S3FFClient {
+  protected readonly api: AxiosInstance;
+
   constructor(
-    protected readonly baseUrl: string,
-    private readonly onProgress: ProgressCallback = () => { /* no-op */ },
+    baseUrl: string,
+    protected readonly onProgress: ProgressCallback = () => { /* no-op */ },
+    apiConfig: AxiosRequestConfig = {},
```

Comment:
I'm not sure whether this should take an `AxiosRequestConfig` or an `AxiosInstance`.

`AxiosRequestConfig` pros:
* Downstreams don't need to install axios just to update the configuration
* It's more obvious syntax (a simple Object) for developers not familiar with axios
* Downstreams already using a pre-configured `AxiosInstance` just need to pass `theAxiosInstance.defaults`, which is not particularly burdensome (but a little non-obvious)

`AxiosInstance` pros:
* It the easiest for downstreams already using a pre-configured `AxiosInstance`
* It's symmetrical with the Python client [taking a pre-configured `requests.Session`](https://github.com/girder/django-s3-file-field/blob/master/python-client/s3_file_field_client/__init__.py#L29)
* Downstream installation of axios should be deduplicated, so it's zero net size

The body of the constructor looks equally simple in both cases, so this is really about what's the best API for downstreams.

I suppose we could take either, and use `instanceof` to do internal detection, but I think a good API should be more opinionated.

Reply:
Do we have an existing use case for needing custom options on this?

I have a slight preference for `AxiosRequestConfig`. In GWC, at the urging of @danlamanna , we chose to go with passing, and in fact directly extending, an `AxiosInstance`. However, that's a different scenario in which the class using it will be used in the library exactly like an axios instance, so it made a lot more sense there. But without knowing why this feature is necessary right now, it's hard for me to choose a side here.

Reply:
I think I have the slightest possible preference for `AxiosInstance`, mainly because of the `instance.defaults` confusion and the symmetry with the python client.

> I suppose we could take either, and use instanceof to do internal detection, but I think a good API should be more opinionated.

I don't think I'm opposed to taking either here. Unless others chime in, this seems so difficult to decide that it might warrant it.

Reply:
If we accept either, let's not do it as some union type. Let's pass it as two different optional keys.